### ISSUE TITLE
Limitation: HTML file must be in the same directory as "demo.html"

### ISSUE THREAD

**Issue by Pylipala** [metafloor/bwip-js#3]
I plan to use bwip-js in my HTML page. So I put all bwip-js files in one library folder called `barcode` and create my HTML files outside of `barcode` folder.
My project structure looks like below:
```
mydemo.html
/barcode
        /bwipp
                code128.js
                code39.js
                qrcode.js
                ...
        /lib
                baropts.js
                canvas.js
                ...
        /proofs
        bwip.js
        demo.html
        freetype.js
        freetype.js.mem
        Inconsolata.otf
```
`mydemo.html` file is copied from `demo.html` and I update all the src properties of script tags, adding `barcode/` prefix. However, it does not work with error `net::ERR_FILE_NOT_FOUND` in browser.

After some digging, I find that `BWIPJS.load` is defined in `canvas.js`. It writes `script` in `document.head` to load javascript files. `BWIPJS.prototype.call` uses `BWIPJS.load` as below
```
		BWIPJS.load('bwipp/' + name + '.js');
```
and every barcode type uses it as below
```
if (!BWIPJS.bwipp["raiseerror"] && BWIPJS.increfs("auspost", "raiseerror")) {
	BWIPJS.load("bwipp/raiseerror.js");
}
if (!BWIPJS.bwipp["renlinear"] && BWIPJS.increfs("auspost", "renlinear")) {
	BWIPJS.load("bwipp/renlinear.js");
}
```
That means the HTML file must be under the same directory of the `bwip` folder to work correctly, i.e. brother of `demo.html`.

I add `barcode/` prefix to `BWIPJS.load` as below
```
BWIPJS.load = function(path) {
	var script = document.createElement('script');
	script.type = 'text/javascript';
	script.src = 'barcode/' + path;
	document.head.appendChild(script);
};
```
and my demo shows barcode well.

But it looks weird to me because now my library folder must be called `barcode` and my HTML's still have to be in the same directory of `barcode` folder.

Do I have another choice or do I have to do it like this? And even I add `barcode/` prefix, there is still error in browser console
```
Failed to load resource: the server responded with a status of 404 (File not found)
freetype.js:15 Uncaught could not load memory initializer freetype.js.mem
```
which is generated code I could not modify.

**Comment by metafloor** [comment]
Unfortunately, these are some of the limitations to running in a browser.  And the emscripten compiled freetype module is tricky to work with.  As you have discovered, it has its own concept of where files live.

Regarding : *and my HTML's still have to be in the same directory of barcode folder*

If you want bwip-js to be located in a sub-directory, then you need to adjust the paths in the `<script>` tags.  The demo is using `src="lib/blah.js"`.  You need to change those to `src="barcode/lib/blah.js"`, in addition to changing the path prefix in the `BWIPJS.load` function.

Regarding : *Uncaught could not load memory initializer freetype.js.mem*

One solution is to add either an HTTP redirect or server-side redirect on your web-server to map, for example, `/foo/bar/freetype.js.mem` to `/foo/bar/barcode/freetype.js.mem`.  The paths in the freetype module are hard-coded at compile time and cannot be changed.

**Comment by metafloor** [comment]
I had some spare time so I played around with your situation and found a poorly documented emscripten option called `memoryInitializerPrefixURL`.  You need to modify your html file that loads freetype.  Look for the `Module` definition and add the option:

```
var Module = {
    memoryInitializerPrefixURL:'barcode/',
    preRun:[ function() { ....
```

BTW, I noticed you have the `proofs` directory.  What version of bwip-js are you using?  The proofs are not needed (I use them for my testing) and were removed several weeks ago.

**Comment by metafloor** [comment]
I have updated `lib/canvas.js` to make the `BWIPJS.load()` method location independent:

```
// Dynamically load the encoders.
BWIPJS.load = function(path) {
    var script = document.createElement('script');
    script.type = 'text/javascript';
    script.src = BWIPJS.load.root + path;
    document.head.appendChild(script);
};

// Determine where this script is located on the server and
// make the encoder loading relative to it.
(function() {
    var scripts = document.getElementsByTagName("script");
    // http://host:port/foo/bar/lib/canvas.js
    // We want root to be everything but lib/canvas.js
    BWIPJS.load.root = scripts[scripts.length-1].src.replace(/lib.canvas.js$/,'');
})();
```

This change is available in v0.12.2.

**Comment by Pylipala** [comment]
Cheers. The `memoryInitializerPrefixURL` options works. New `lib/canvas.js` also works.

Regarding: *I noticed you have the proofs directory. What version of bwip-js are you using?*

I am using the `gh-pages` branch.
